### DISCUSSION TITLE
Transit and Customer Traffic on Trunk port

### DISCUSSION THREAD

**Discussion by smallsam** [akvorado/akvorado#318] [Q&A]
I have a scenario where we are tapping sflow on an edge switch interface to a partner provider. However this private link handles both Transit for us and downstream customers on different 802.1q vlans (sadly the switch only can export sflow at the physical interface).

Having a read of: https://github.com/akvorado/akvorado/blob/main/inlet/flow/decoder/helper.go suggests that you don't handle vlans at all from the sflow/netflow messages and I'm wondering how best to approach this.

My thought is to coalesce the idea of a vlan with that of an interface, such that a new input vlan can be considered a new interface - this would allow different vlans to be annotated with different "providers" or "connectivity" as necessary. This metadata could be provided in the config but it would be elegant if it too could be derived from SNMP. I was considering adding a sub-interface on the switch for each vlan that requires new metadata (ifAlias) to be assigned rather than querying a global vlan table. This would allow much of the infrastructure for handing this metadata to be retained. I'll have to check how the interface-subinterface relationship is mapped in SNMP, that's a part I haven't walked through yet.

This is probably the most ideal option for me to move forward with but it'll take a bit of effort to make the required patch (I'm no go programmer), so I'm exploring other options including frontending with vflow, goflow2 or pmacct to filter first, this would certainly be less elegant.

Any thoughts from the community would be great!

**Comment by vincentbernat** [comment]
Getting relationship between subinterfaces is easy with SNMP. That's the `IF-MIB::ifStackTable`.

Maybe check first with Wireshark if you don't directly get the ifIndex of each VLAN you receive through sflow?

**Reply by smallsam** [reply]
Good idea, I did setup the subinterface and would have expected to see the new descriptions/interface classification flow through to akvorado had it have been passing the subinterface ifindex in the flows. 

The ifIndex appears to remain as the physical port, I think I only get the vlan_id in the sampled packet:
![image](https://user-images.githubusercontent.com/208799/206601363-f8b874c5-bb95-4404-9b43-397896053355.png)

Which I assume ends up as vlan_id here: https://github.com/netsampler/goflow2/blob/332655476b0a1025647e9eca087920e46cb7b7d4/pb/flow.proto#L61

Having an input vlan and output vlan feels like it would have aligned more neatly with interfaces, I assume that's not populated here because all the ports I'm exporting are trunks and so the vlan is only in the packet header.

**Reply by smallsam** [reply]
Digging into this a little deeper, the switch in question is a Dlink DXS-5000-54S whose current OS doesn't implement IF-MIB::ifStackTable. It's probably a little too much to hope that's implemented in the most recent software.

**Reply by vincentbernat** [reply]
This is not surprising, many vendors don't implement it. Sometimes, there are other ways to get the information, notably by using the vendor MIBs. To be honest, this would require a huge amount of work for a corner use case. The VLAN ID could be exposed to the interface classifier rules, so that you could do the classification. We could also allow rules to update the interface name and descriptions.

**Reply by smallsam** [reply]
Just trying to think how to do this in an extensible way (in case there are use cases to inspect other flow fields for classification and also trying to avoid additional SNMP work I think we can do the following.

 * Add concept of flow classifiers, this makes updating cache behaviour more easy to understand.
 * Add Flow.VlanId (and potentially other flow fields) to flow classifiers. This could allow for rules to directly set connectity and provider based on flow field values.
 * In lieu of rules to update names, extend regex classifiers to allow specification of the test string and substitution of fields into the regex:
ClassifyConnectivityRegex(Interface.Description, "^${Flow.VlanId}:(transit|pni|ppni|ix):? ", "$1")
This would allow encoding of vlan specific connectivity and provider info (as per this discussion) into the interface description (ifAlias) so that you do not need new classifier rules added when a new provider is introduced.

I'd like to move ahead with this in a way that you'd be happy to merge in so let me know if this sounds like a good direction?

**Reply by vincentbernat** [reply]
I don't understand what you mean by flow classifiers. We have interface classifiers and exporter classifiers. I am OK with adding Vlan ID as an input of interface classifiers to let you do rules for that. As for allowing substitutions inside regex, I don't understand what you want to achieve. A complete example of an interface classifier may help.

**Comment by smallsam** [comment]
Functionality to meet this requirement is being developed in https://github.com/akvorado/akvorado/pull/384.
